### PR TITLE
tensor disposal in data loader

### PR DESCRIPTION
**Pull request by yueyinqiu**
Main changes:

- add `disposeBatch` option to indicate whether to dispose the collated tensors automatically
    - otherwise data loaders cannot work with some linq methods like `First`, because the returned tensors would be disposed
    - currently we are still using a list. if we allows ["custom" dispose scopes](https://github.com/dotnet/TorchSharp/pull/1304#issuecomment-2088234426), we could switch to this
- rewrite the enumerator

---

Other changes:

- collate functions will now dispose the intermediate tensors, rather than wait for the next iteration
- add `disposeDataset` option
- keep the aliases of the tensors in `TensorDataset`

---

Perhaps we need a document about custom datasets and dataloaders.

### CODE REVIEW

**Comment by yueyinqiu** [comment]
The tensors got from `dataset.GetTensor` is not disposed. This could cause memory leaks when using some datasets, but might also cause a unexpected disposal. I would check that and use `alias` when returning those tensors.

Perhaps we need a document about custom datasets and dataloaders.

**Comment by yueyinqiu** [comment]
data loader will also dispose the given dataset. so the name `autoDispose` seems to be not that clear.

**Comment by yueyinqiu** [comment]
Found that the shuffle is disposed after the first iteration...

```csharp
using System.Collections;
using TorchSharp;

var dataset = torch.utils.data.TensorDataset(torch.zeros([3]));
var dataLoader = torch.utils.data.DataLoader(dataset, 1, new MyShuffler());
foreach (var _ in dataLoader) ; // ObjectDisposedException

class MyShuffler : IEnumerable<long>
{
    public IEnumerator<long> GetEnumerator() => new ShufflerEnumerator();
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    class ShufflerEnumerator : IEnumerator<long>
    {
        bool disposed = false;
        public long Current
        {
            get
            {
                if (disposed)
                    throw new ObjectDisposedException("");
                return 1;
            }
        }
        object IEnumerator.Current => Current;
        public void Dispose()
        {
            disposed = true;
        }
        public bool MoveNext() => true;
        public void Reset() { }
    }
}
```

**Comment by yueyinqiu** [comment]
Hmmm... `drop_last` is also ignored:

```csharp
using TorchSharp;

var dataset = torch.utils.data.TensorDataset(torch.tensor(new long[] { 1, 2, 3 }));
Console.WriteLine(dataset.Count); // 3
var dataLoader = torch.utils.data.DataLoader(dataset, 2, drop_last: true);
Console.WriteLine(dataLoader.Count); // 1
Console.WriteLine(dataLoader.Count()); // 2
```

**Comment by NiklasGustafsson** [comment]
Tests related to DataLoader are failing.

**Comment by yueyinqiu** [comment]
Ah... I forgot to check whether it is coming to an end...
